pkg/internal/ebpf/tracer_linux.go

```diff
@@ -238,64 +261,504 @@ func (pt *ProcessTracer) Init() error {
+
+	if err != nil {
+		return fmt.Errorf("failed to open executable: %w", err)
+	}
+
+	closers, err := i.instrumentProbes(exe, goProbes)
+
+	if err != nil {
+		printVerifierErrorInfo(err)
+		return fmt.Errorf("failed to attach go probes: %w", err)
+	}
+
+	i.addModule(ie.FileInfo.Ino)
+	pt.recordInstrumentedBin(ie.FileInfo.Ino, closers)
+	pt.Instrumentables[ie.FileInfo.Ino] = i
+
+	return nil
+}
+
+// assigns uprobes to binaries, or to the main binary in case static linkage
+// is suspected
+func resolveBinProbes(uProbes uProbes, binaries []processBinary) {
+	if len(binaries) == 0 {
+		return
+	}
+
+	// the process main binary is always first
+	mainBinary := &binaries[0]
+
+	// if a shared library is not present, reassign its uprobes to the main
+	// binary in case they are statically linked
+	for fileName, uProbe := range uProbes {
+		binary := findIf(binaries, func(p processBinary) bool {
+			return strings.HasPrefix(filepath.Base(p.Path), fileName)
+		})
+
+		if binary == nil {
+			binary = mainBinary
+		}
+
+		for k, v := range uProbe {
+			binary.UProbes[k] = v
 		}
+	}
+}
+
+func (pt *ProcessTracer) attachUProbes(ie *Instrumentable, i *instrumenter) error {
+	uProbes := pt.gatherUProbes()
+
+	if len(uProbes) == 0 {
+		return nil
+	}
+
+	binaries, err := gatherProcessBinaries(ie)
+
+	if err != nil {
+		return err
+	}
+
+	if len(binaries) == 0 {
+		return fmt.Errorf("no binaries to instrument")
+	}
+
+	resolveBinProbes(uProbes, binaries)
 
-		// Uprobes to be used for native module instrumentation points
-		if err := i.uprobes(ie.FileInfo.Pid, p); err != nil {
+	binAbsPath := func(b *processBinary) string {
+		return fmt.Sprintf("/proc/%d/root%s", ie.FileInfo.Pid, b.Path)
+	}
+
+	for _, binary := range binaries {
+		if len(binary.UProbes) == 0 {
+			continue
+		}
+
+		if pt.alreadyInstrumentedBin(binary.Ino) {
+			pt.log.Debug("module already instrumented by other processes, incrementing reference count",
+				"path", binary.Path, "ino", binary.Ino)
+
+			i.addModule(binary.Ino)
+			pt.addInstrumentedBinRef(binary.Ino)
+			continue
+		}
+
+		absPath := binAbsPath(&binary)
+
+		targetExe, err := link.OpenExecutable(absPath)
+
+		if err != nil {
+			return fmt.Errorf("failed to open binary %s: %w", absPath, err)
+		}
+
+		if err := resolveUProbeOffsets(absPath, binary.UProbes); err != nil {
+			return fmt.Errorf("failed to resolve uprobe offsets: %w", err)
+		}
+
+		closers, err := i.instrumentProbes(targetExe, binary.UProbes)
+
+		if err != nil {
 			printVerifierErrorInfo(err)
-			return err
+			return fmt.Errorf("failed to attach uprobes: %w", err)
 		}
+
+		i.addModule(binary.Ino)
+		pt.recordInstrumentedBin(binary.Ino, closers)
 	}
 
-	pt.Instrumentables[ie.FileInfo.Ino] = &i
+	pt.Instrumentables[ie.FileInfo.Ino] = i
+
+	return nil
+}
+
+func (pt *ProcessTracer) NewExecutable(ie *Instrumentable) error {
+	i := instrumenter{
+		modules: map[uint64]struct{}{},
+	}
+
+	if pt.Type == Go {
```

Comment:
This is missing RegisterOffsets, where is register offsets called now?

Reply:
I see, they are now in Goprobes, but that's not great, they must be called for all tracers, not just for Go. GPU tracer relies on them and in the future we might have offsets that we need pass down for the other tracers based on things that are hardcoded for the generic tracer at the moment.